File: Source/Extensions/Tessa.Extensions.Default.Client/Views/AdvancedFilterViewDialogManager.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Platform.Formatting;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Views.Parameters;
using Tessa.Views;
using Tessa.Views.Metadata;
using Tessa.Views.Parser;

namespace Tessa.Extensions.Default.Client.Views
{
    /// <inheritdoc cref="IAdvancedFilterViewDialogManager"/>
    public class AdvancedFilterViewDialogManager :
        IAdvancedFilterViewDialogManager
    {
        #region Fields

        private readonly CreateDialogFormFuncAsync createDialogFormFuncAsync;

        private readonly IAdvancedCardDialogManager advancedCardDialogManager;

        #endregion

        #region Constructors

        /// <summary>
        /// Инициализирует новый экземпляр класса.
        /// </summary>
        /// <param name="createDialogFormFuncAsync"><inheritdoc cref="CreateDialogFormFuncAsync" path="/summary"/></param>
        /// <param name="advancedCardDialogManager"><inheritdoc cref="IAdvancedCardDialogManager" path="/summary"/></param>
        public AdvancedFilterViewDialogManager(
            CreateDialogFormFuncAsync createDialogFormFuncAsync,
            IAdvancedCardDialogManager advancedCardDialogManager)
        {
            this.createDialogFormFuncAsync = NotNullOrThrow(createDialogFormFuncAsync);
            this.advancedCardDialogManager = NotNullOrThrow(advancedCardDialogManager);
        }

        #endregion

        #region IAdvancedViewParametersDialogManager Members

        /// <inheritdoc/>
        public virtual async Task OpenAsync(
            FilterViewDialogDescriptor descriptor,
            IViewParameters parameters,
            CancellationToken cancellationToken = default)
        {
            ThrowIfNull(descriptor);
            ThrowIfNull(parameters);

            var dialogCardModel = await this.CreateDialogCardModelAsync(
                descriptor.DialogName,
                descriptor.FormAlias,
                cancellationToken);

            await this.ShowDialogAsync(
                descriptor,
                parameters,
                dialogCardModel,
                cancellationToken);
        }

        #endregion

        #region Protected Methods

        /// <summary>
        /// Создаёт модель карточки диалога, содержащего параметры представления.
        /// </summary>
        /// <param name="dialogName">Имя типа диалога.</param>
        /// <param name="formAlias">Алиас формы диалога или <see langword="null"/>, если требуется создать форму для первой вкладки типа диалога.</param>
        /// <param name="cancellationToken">Объект, посредством которого можно отменить асинхронную задачу.</param>
        /// <returns>Модель карточки диалога.</returns>
        /// <exception cref="InvalidOperationException">При создании модели карточки диалога произошла ошибка.</exception>
        protected async ValueTask<ICardModel> CreateDialogCardModelAsync(
            string dialogName,
            string? formAlias,
            CancellationToken cancellationToken = default)
        {
            (var form, var cardModel) = await this.createDialogFormFuncAsync(
                dialogName,
                formAlias,
                modifyModelAsync: static (newCardModel, _) =>
                {
                    newCardModel.Card.Version = 1;
                    newCardModel.Flags |= CardModelFlags.IgnoreChanges;

                    return ValueTask.CompletedTask;
                },
                cancellationToken: cancellationToken);

            if (form is null
                || cardModel is null)
            {
                throw new InvalidOperationException(
                    $"Failed to create dialog. Dialog name: \"{dialogName}\". Form alias: \"{FormattingHelper.FormatNullable(formAlias)}\".");
            }

            cardModel.MainForm = form;

            return cardModel;
        }

        /// <summary>
        /// Отображает диалог, содержащий параметры фильтрации представления.
        /// </summary>
        /// <param name="descriptor"><inheritdoc cref="FilterViewDialogDescriptor" path="/summary"/></param>
        /// <param name="parameters"><inheritdoc cref="IViewParameters" path="/summary"/></param>
        /// <param name="dialogCardModel">Модель карточки диалога.</param>
        /// <param name="cancellationToken">Объект, посредством которого можно отменить асинхронную задачу.</param>
        /// <returns>Асинхронная задача.</returns>
        protected virtual async Task ShowDialogAsync(
            FilterViewDialogDescriptor descriptor,
            IViewParameters parameters,
            ICardModel dialogCardModel,
            CancellationToken cancellationToken = default)
        {
            try
            {
                var isApplied = false;
                var context = await this.advancedCardDialogManager.ShowCardAsync(
                    dialogCardModel,
                    prepareEditorActionAsync: (editor, _) =>
                    {
                        editor.StatusBarIsVisible = false;

                        FillFields(
                            parameters,
                            editor.CardModel.Card,
                            descriptor.ParametersMapping);

                        editor.Toolbar.Actions.Clear();
                        editor.BottomToolbar.Actions.Clear();
                        editor.BottomDialogButtons.Clear();

                        editor.BottomDialogButtons.Add(
                            new UIButton(
                                "$UI_Common_OK",
                                async _ =>
                                {
                                    var editor = UIContext.Current.CardEditor;

                                    if (editor is null
                                        || editor.OperationInProgress)
                                    {
                                        return;
                                    }

                                    isApplied = true;

                                    await editor.CloseAsync(
                                        cancellationToken: CancellationToken.None);
                                },
                                isDefault: true));

                        editor.BottomDialogButtons.Add(
                            new UIButton(
                                "$UI_Common_Cancel",
                                async static _ =>
                                {
                                    var editor = UIContext.Current.CardEditor;

                                    if (editor is null
                                        || editor.OperationInProgress)
                                    {
                                        return;
                                    }

                                    await editor.CloseAsync(
                                        cancellationToken: CancellationToken.None);
                                },
                                isCancel: true));

                        return ValueTask.FromResult(true);
                    },
                    options: new OpenCardOptions
                    {
                        DisplayValue = "$Views_FilterDialog_Caption",
                        WithDialogWallpaper = false,
                        DialogWindowModifierAction = static window =>
                        {
                            window.MaxWidth = 800;
                            window.SizeToContent = SizeToContent.Height;
                        }
                    },
                    cancellationToken: cancellationToken);

                if (isApplied)
                {
                    using (parameters.SuspendChangesNotification())
                    {
                        parameters.Clear();
                        CreateParameters(
                            parameters,
                            context.CardEditor.CardModel.Card,
                            descriptor.ParametersMapping);
                    }
                }
            }
            catch (NotSupportedException)
            {
                // Используется FakeUIHost, например, мы открыты в TessaAdmin, игнорируем ошибку.
            }
        }

        /// <summary>
        /// Заполняет поля карточки, данными параметров запроса к представлению.
        /// </summary>
        /// <param name="parameters">Список параметров представления.</param>
        /// <param name="card">Карточка, содержащая параметры.</param>
        /// <param name="parameterMappings">Коллекция, содержащая информацию о связи параметров представления и полей карточки.</param>
        protected static void FillFields(
            IViewParameters parameters,
            Card card,
            IReadOnlyCollection<ParameterMapping> parameterMappings)
        {
            var sections = card.Sections;

            foreach (var parameterMapping in parameterMappings)
            {
                FillField(
                    parameters,
                    sections,
                    parameterMapping);
            }
        }

        /// <summary>
        /// Заполняет поле <see cref="ParameterMapping.ValueFieldName"/> в секции <see cref="ParameterMapping.ValueSectionName"/>, содержащее параметр фильтрации представления <see cref="ParameterMapping.Alias"/>.
        /// </summary>
        /// <param name="parameters">Список параметров представления.</param>
        /// <param name="sections">Секции, содержащиеся в карточке диалога с параметрами представления.</param>
        /// <param name="parameterMapping"><inheritdoc cref="ParameterMapping" path="/summary"/></param>
        protected static void FillField(
            IViewParameters parameters,
            IReadOnlyDictionary<string, CardSection> sections,
            ParameterMapping parameterMapping)
        {
            var parameter = parameters.FirstOrDefault(
                i =>
                    ParserNames.IsEquals(
                        i.Metadata.Alias,
                        parameterMapping.Alias));

            if (parameter is null)
            {
                return;
            }

            var value = parameter
                .CriteriaValues
                .FirstOrDefault()
                ?.Values
                .FirstOrDefault();

            if (value is null)
            {
                return;
            }

            if (!sections.TryGetValue(parameterMapping.ValueSectionName, out var valueSection))
            {
                return;
            }

            valueSection.Fields[parameterMapping.ValueFieldName] = value.Value;

            if (string.IsNullOrEmpty(parameterMapping.DisplayValueSectionName)
                || !sections.TryGetValue(parameterMapping.DisplayValueSectionName, out var displayValueSection))
            {
                return;
            }

            if (!string.IsNullOrEmpty(parameterMapping.DisplayValueFieldName))
            {
                displayValueSection.Fields[parameterMapping.DisplayValueFieldName] = value.Text;
            }
        }

        /// <summary>
        /// Создаёт параметры запроса к представлению.
        /// </summary>
        /// <param name="parameters">Список параметров представления.</param>
        /// <param name="card">Карточка, содержащая параметры.</param>
        /// <param name="parameterMappings">Коллекция, содержащая информацию о связи параметров представления и полей карточки.</param>
        protected static void CreateParameters(
            IViewParameters parameters,
            Card card,
            IReadOnlyCollection<ParameterMapping> parameterMappings)
        {
            var sections = card.Sections;

            foreach (var parameterMapping in parameterMappings)
            {
                AddParameterIfValueNotEmpty(
                    parameters,
                    sections,
                    parameterMapping);
            }
        }

        /// <summary>
        /// Добавляет параметр в запрос к представлению, если поле <see cref="ParameterMapping.ValueFieldName"/> в секции <see cref="ParameterMapping.ValueSectionName"/> содержит данные.
        /// </summary>
        /// <param name="parameters">Список параметров представления.</param>
        /// <param name="sections">Секции, содержащиеся в карточке диалога с параметрами представления.</param>
        /// <param name="parameterMapping"><inheritdoc cref="ParameterMapping" path="/summary"/></param>
        protected static void AddParameterIfValueNotEmpty(
            IViewParameters parameters,
            IReadOnlyDictionary<string, CardSection> sections,
            ParameterMapping parameterMapping)
        {
            if (!sections.TryGetValue(parameterMapping.ValueSectionName, out var valueSection))
            {
                return;
            }

            if (!valueSection.Fields.TryGetValue(parameterMapping.ValueFieldName, out var valueField)
                || valueField is null)
            {
                return;
            }

            var parameterMetadata = parameters.Metadata.FirstOrDefault(
                i =>
                    ParserNames.IsEquals(
                        i.Alias,
                        parameterMapping.Alias));

            if (parameterMetadata is null)
            {
                return;
            }

            string? displayValue;

            if (!string.IsNullOrEmpty(parameterMapping.DisplayValueSectionName)
                && !string.IsNullOrEmpty(parameterMapping.DisplayValueFieldName)
                && sections.TryGetValue(
                    parameterMapping.DisplayValueSectionName,
                    out var displayValueSection)
                && displayValueSection.RawFields.TryGetValue(
                    parameterMapping.DisplayValueFieldName,
                    out var displayValueObj)
                && displayValueObj is not null)
            {
                displayValue = displayValueObj.ToString();
            }
            else
            {
                displayValue = valueField.ToString();
            }

            var requestParameter = new RequestParameterBuilder()
                .WithMetadata(parameterMetadata)
                .AddCriteria(
                    parameterMapping.CriteriaOperator
                        ?? parameterMetadata.GetDefaultCriteria(),
                    displayValue,
                    valueField)
                .AsRequestParameter();

            parameters.Add(requestParameter);
        }

        #endregion
    }
}

File: WebClient SDK/src/default/views/advancedFilterViewDialogManager.ts
import { AdvancedCardDialogManager, CardModelFlags, ICardModel } from 'tessa/ui/cards';
import { Card, CardSection } from 'tessa/cards';
import { UIButton, UIContext, createDialogForm } from 'tessa/ui';

import { DotNetType } from 'tessa/platform';
import { FilterViewDialogDescriptor } from './filterViewDialogDescriptor';
import { IViewParameters } from 'tessa/ui/views/parameters';
import { MapStorage } from 'tessa/platform/storage';
import { ParameterMapping } from './parameterMapping';
import { RequestParameter } from 'tessa/views/metadata';
import { RequestParameterBuilder } from 'tessa/views';
import { runInAction } from 'mobx';

/**
 * Объект, предоставляющий методы для открытия модального диалога с параметрами фильтрации представления.
 */
export interface IAdvancedFilterViewDialogManager {
  /**
   * Открывает диалог с параметрами фильтрации представления.
   * @param descriptor {@link FilterViewDialogDescriptor}
   * @param parameters {@link IViewParameters}
   */
  open(descriptor: FilterViewDialogDescriptor, parameters: IViewParameters): Promise<void>;
}

export class AdvancedFilterViewDialogManager implements IAdvancedFilterViewDialogManager {
  //#region ctor

  private constructor() {}

  //#endregion

  //#region instance
  private static _instance: AdvancedFilterViewDialogManager;

  /**
   * Экземпляр этого объекта.
   */
  public static get instance(): AdvancedFilterViewDialogManager {
    if (!AdvancedFilterViewDialogManager._instance) {
      AdvancedFilterViewDialogManager._instance = new AdvancedFilterViewDialogManager();
    }

    return AdvancedFilterViewDialogManager._instance;
  }

  //#endregion

  //#region IAdvancedFilterViewDialogManager members

  async open(descriptor: FilterViewDialogDescriptor, parameters: IViewParameters): Promise<void> {
    const dialogCardModel = await this.createDialogCardModel(
      descriptor.dialogName,
      descriptor.formAlias
    );

    await this.showDialog(descriptor, parameters, dialogCardModel);
  }

  //#endregion

  //#region protected methods

  /**
   * Создаёт модель карточки диалога, содержащего параметры представления.
   * @param dialogName Имя типа диалога.
   * @param formAlias Алиас формы диалога или undefined, если требуется создать форму для первой вкладки типа диалога.
   * @returns Модель карточки диалога.
   */
  protected async createDialogCardModel(
    dialogName: string,
    formAlias: string | undefined
  ): Promise<ICardModel> {
    const result = await createDialogForm(
      'CarViewParameters',
      'MainTab',
      undefined,
      undefined,
      async (newCardModel: ICardModel) => {
        newCardModel.card.version = 1;
        newCardModel.flags |= CardModelFlags.IgnoreChanges;
      }
    );
    if (!result) {
      throw new Error(
        `Failed to create dialog. Dialog name: "${dialogName}". Form alias: "${
          formAlias ?? 'undefined'
        }".`
      );
    }

    const [form, cardModel] = result;
    cardModel.mainForm = form;

    return cardModel;
  }

  /**
   * Отображает диалог, содержащий параметры фильтрации представления.
   * @param descriptor {@link FilterViewDialogDescriptor}
   * @param parameters {@link IViewParameters}
   * @param dialogCardModel Модель карточки диалога.
   */
  protected async showDialog(
    descriptor: FilterViewDialogDescriptor,
    parameters: IViewParameters,
    dialogCardModel: ICardModel
  ): Promise<void> {
    await AdvancedCardDialogManager.instance.showCardModel({
      cardModel: dialogCardModel,
      displayValue: '$Views_FilterDialog_Caption',
      dialogOptions: {
        withDialogWallpaper: false,
        dialogAutoSize: true
      },
      prepareEditorAction: editor => {
        editor.statusBarIsVisible = false;

        AdvancedFilterViewDialogManager.fillFields(
          parameters,
          editor.cardModel!.card,
          descriptor.parametersMapping
        );

        editor.toolbar.clearItems();
        editor.bottomToolbar.clearItems();
        editor.bottomDialogButtons.length = 0;

        editor.bottomDialogButtons.push(
          UIButton.create({
            caption: '$UI_Common_OK',
            buttonAction: async () => {
              const editor = UIContext.current.cardEditor;

              if (!editor || editor.operationInProgress) {
                return;
              }

              const params = AdvancedFilterViewDialogManager.createParameters(
                parameters,
                editor.cardModel!.card,
                descriptor.parametersMapping
              );
              runInAction(() => {
                parameters.clear();
                parameters.addParameters(...params);
              });

              await editor.close();
            }
          })
        );

        editor.bottomDialogButtons.push(
          UIButton.create({
            caption: '$UI_Common_Cancel',
            buttonAction: async () => {
              const editor = UIContext.current.cardEditor;

              if (!editor || editor.operationInProgress) {
                return;
              }

              await editor.close();
            }
          })
        );

        return true;
      }
    });
  }

  /**
   * Заполняет поля карточки, данными параметров запроса к представлению.
   * @param parameters Список параметров представления.
   * @param card Карточка, содержащая параметры.
   * @param parameterMappings Коллекция, содержащая информацию о связи параметров представления и полей карточки.
   */
  protected static fillFields(
    parameters: IViewParameters,
    card: Card,
    parameterMappings: ReadonlyArray<ParameterMapping>
  ): void {
    const sections = card.sections;

    for (const parameterMapping of parameterMappings) {
      AdvancedFilterViewDialogManager.fillField(parameters, sections, parameterMapping);
    }
  }

  /**
   * Заполняет поле {@link ParameterMapping.valueFieldName} в секции {@link ParameterMapping.valueSectionName}, содержащее параметр фильтрации представления {@link ParameterMapping.alias}.
   * @param parameters Список параметров представления.
   * @param sections Секции, содержащиеся в карточке диалога с параметрами представления.
   * @param parameterMapping {@link ParameterMapping}
   */
  protected static fillField(
    parameters: IViewParameters,
    sections: MapStorage<CardSection>,
    parameterMapping: ParameterMapping
  ): void {
    const parameter = parameters.parameters.find(x => x.metadata?.alias === parameterMapping.alias);

    if (!parameter) {
      return;
    }

    const value = parameter.criteriaValues[0]?.values[0];
    if (value == undefined) {
      return;
    }

    const valueSection = sections.get(parameterMapping.valueSectionName);

    if (!valueSection) {
      return;
    }

    valueSection.fields.set(
      parameterMapping.valueFieldName,
      value.value,
      parameter.metadata?.schemeType?.dotNetType ?? DotNetType.String
    );

    if (!parameterMapping.displayValueSectionName || !parameterMapping.displayValueFieldName) {
      return;
    }

    const displayValueSection = sections.get(parameterMapping.displayValueSectionName);

    if (!displayValueSection) {
      return;
    }

    displayValueSection.fields.set(
      parameterMapping.displayValueFieldName,
      value.text,
      DotNetType.String
    );
  }

  /**
   * Создаёт параметры запроса к представлению.
   * @param parameters Список параметров представления.
   * @param card Карточка, содержащая параметры.
   * @param parameterMappings >Коллекция, содержащая информацию о связи параметров представления и полей карточки.
   * @returns Список параметров, передаваемых в запросе к представлению.
   */
  protected static createParameters(
    parameters: IViewParameters,
    card: Card,
    parameterMappings: ReadonlyArray<ParameterMapping>
  ): RequestParameter[] {
    const params: RequestParameter[] = [];
    const sections = card.sections;

    for (const parameterMapping of parameterMappings) {
      AdvancedFilterViewDialogManager.addParameterIfValueNotEmpty(
        params,
        parameters,
        sections,
        parameterMapping
      );
    }

    return params;
  }

  /**
   * Добавляет параметр в запрос к представлению, если поле {@link ParameterMapping.valueFieldName} в секции{@link ParameterMapping.valueSectionName} содержит данные.
   * @param paramsStorage Список параметров, передаваемых в запросе к представлению.
   * @param parameters Список параметров представления.
   * @param sections Секции, содержащиеся в карточке диалога с параметрами представления.
   * @param parameterMapping {@link ParameterMapping}
   */
  protected static addParameterIfValueNotEmpty(
    paramsStorage: RequestParameter[],
    parameters: IViewParameters,
    sections: MapStorage<CardSection>,
    parameterMapping: ParameterMapping
  ): void {
    const valueSection = sections.get(parameterMapping.valueSectionName);

    if (!valueSection) {
      return;
    }

    const fieldValue = valueSection.fields.tryGet(parameterMapping.valueFieldName);

    if (fieldValue == undefined) {
      return;
    }

    const parameterMetadata = parameters.metadata.find(x => x.alias === parameterMapping.alias);

    if (!parameterMetadata) {
      return;
    }

    let displayValue = fieldValue.toString();
    if (parameterMapping.displayValueSectionName && parameterMapping.displayValueFieldName) {
      const displayValueSection = sections.get(parameterMapping.displayValueSectionName);

      if (displayValueSection) {
        const displayValueField = displayValueSection.fields.tryGet(
          parameterMapping.displayValueFieldName
        );

        if (displayValueField != undefined) {
          displayValue = displayValueField.toString();
        }
      }
    }

    const requestParameter = new RequestParameterBuilder()
      .withMetadata(parameterMetadata)
      .addCriteria(
        parameterMapping.criteriaOperator ?? parameterMetadata.getDefaultCriteria(),
        displayValue,
        fieldValue
      )
      .asRequestParameter();

    paramsStorage.push(requestParameter);
  }

  //#endregion
}
